src/ooxml/java/org/apache/poi/xddf/usermodel/chart/XDDFChart.java

```diff
@@ -346,5 +381,268 @@ private void parseAxes() {
             axes.add(new XDDFValueAxis(valAx));
         }
     }
-
+    /**
+     * method to create relationship with embedded part
+     * for example writing xlsx file stream into output stream
+     * @param chartRelation relationship object
+     * @param chartFactory ChartFactory object
+     * @param chartIndex index used to suffix on file
+     * @return return relation part which used to write relation in .rels file and get relation id
+     * @since POI 4.0.0
+     */
+    public PackageRelationship createRelationshipInChart(POIXMLRelation chartRelation,POIXMLFactory chartFactory,int chartIndex) {
+        documentPart = createRelationship(chartRelation, chartFactory, chartIndex, true).getDocumentPart();
+        return this.addRelation(null, chartRelation, documentPart).getRelationship();
+    }
+    
+    /**
+     * if embedded part was null then create new part
+     * @param chartRelation chart relation object
+     * @param chartWorkbookRelation chart workbook relation object
+     * @param chartFactory factory object of POIXMLFactory (XWPFFactory/XSLFFactory)
+     * @return return the new package part
+     * @throws InvalidFormatException
+     * @since POI 4.0.0
+     */
+    private PackagePart createWorksheetPart(POIXMLRelation chartRelation,POIXMLRelation chartWorkbookRelation,POIXMLFactory chartFactory) throws InvalidFormatException {
+        PackageRelationship xlsx = createRelationshipInChart(XSLFRelation.WORKBOOK_RELATIONSHIP, XSLFFactory.getInstance(), chartIndex);
+        this.setExternalId(xlsx.getId());
+        return getTargetPart(xlsx);
+    }
+    
+    /**
+     * this method write the XSSFWorkbook object data into embedded excel file
+     * @param workbook XSSFworkbook object
+     * @throws IOException
+     * @throws InvalidFormatException
+     * @since POI 4.0.0
+     */
+    public void saveWorkbook(XSSFWorkbook workbook) throws IOException, InvalidFormatException {
+        PackagePart worksheetPart = getWorksheetPart();
+        if (worksheetPart == null) {
+            POIXMLRelation chartRelation = null;
+            POIXMLRelation chartWorkbookRelation = null;
+            POIXMLFactory chartFactory = null;
+            if (this instanceof XSLFChart)
```

Comment:
I remember that I did fix that code smell a few weeks ago. What branch is this code based on?